test/e2e/test/step.go

```diff
@@ -32,23 +51,165 @@ func (l StepList) WithStep(testStep Step) StepList {
+			logf.Log.Error(errors.New("test failure"), "continuing with additional tests")
 			if ts.OnFailure != nil {
 				ts.OnFailure()
 			}
-			t.FailNow()
+			ctx := Ctx()
+			defer func() {
+				if err := deleteElasticResources(); err != nil {
+					logf.Log.Error(err, "while deleting elastic resources")
+				}
+			}()
+			if canRunDiagnostics(ctx) {
+				gcpCredentialsPath = ctx.GCPCredentialsPath
+				once.Do(initGSUtil)
+				logf.Log.Info("running eck-diagnostics job")
+				runECKDiagnostics(ctx, t.Name(), ts)
+				uploadDiagnosticsArtifacts()
+			}
+		}
+	}
+}
+
+// canRunDiagnostics will determine if this e2e test run has the ability to run eck-diagnostics after
+// each test failure, which includes uploading the resulting zip file to a GS bucket. If the job name
+// is set (not empty) and the google credentials file exists, then we should be able to run diagnostics.
+func canRunDiagnostics(ctx Context) bool {
+	if _, err := os.Stat(ctx.GCPCredentialsPath); err != nil && errors.Is(err, fs.ErrNotExist) {
+		return false
+	} else if err != nil {
+		log.Error(err, "while checking for existence of %s", ctx.GCPCredentialsPath)
+		return false
+	}
+	if ctx.JobName == "" {
+		return false
+	}
+	return true
+}
+
+func initGSUtil() {
+	cmd := exec.Command("gcloud", "auth", "activate-service-account", fmt.Sprintf("--key-file=%s", gcpCredentialsPath)) //nolint:gosec
+	setupStdOutErr(cmd)
+	cmd.Env = ensureTmpHomeEnv(cmd.Environ())
+	if err := cmd.Run(); err != nil {
+		log.Error(err, fmt.Sprintf("while initializing gsutil: %s", err))
+	}
+}
+
+func runECKDiagnostics(ctx Context, testName string, step Step) {
+	otherNS := append([]string{ctx.E2ENamespace}, ctx.Operator.ManagedNamespaces...)
+	// The following appends the test, and it's sub-test names together with a '-'.
+	// Example: For TestAutoscalingLegacy/Secrets_should_eventually_be_created:
+	// testName: TestAutoscalingLegacy, step.Name: Secrets_should_eventually_be_created
+	fullTestName := fmt.Sprintf("%s-%s", testName, step.Name)
+	// Convert any spaces to "_", and "/" to "-" in the test name.
+	normalizedTestName := strings.ReplaceAll(strings.ReplaceAll(fullTestName, " ", "_"), "/", "-")
+	cmd := exec.Command("eck-diagnostics", "--output-directory", "/tmp", "-n", fmt.Sprintf("eck-diagnostic-%s.zip", normalizedTestName), "-o", ctx.Operator.Namespace, "-r", strings.Join(otherNS, ","), "--run-agent-diagnostics") //nolint:gosec
+	setupStdOutErr(cmd)
+	// If /tmp is not set as HOME, the k8s client can't cache responses,
+	// and a single run of diagnostics is incredibly slow.
+	cmd.Env = ensureTmpHomeEnv(cmd.Environ())
+	if err := cmd.Run(); err != nil {
+		log.Error(err, fmt.Sprintf("while running eck-diagnostics: %s", err))
+	}
+}
+
+func uploadDiagnosticsArtifacts() {
+	ctx := Ctx()
+	cmd := exec.Command("gsutil", "cp", "/tmp/*.zip", fmt.Sprintf("gs://eck-e2e-buildkite-artifacts/jobs/%s/%s/", ctx.JobName, ctx.BuildNumber)) //nolint:gosec
```

Comment:
Should this bucket be configuration rather than being hard-coded? Also who is cleaning the artefacts up once they have been copied out of this storage bucket

Reply:
Bucket name is now a configuration/flag, and the bucket is cleaned up in the bash script.